src/main/scala/com/hortonworks/spark/sql/hive/llap/HS2JDBCWrapper.scala

```diff
@@ -166,12 +168,38 @@ class JDBCWrapper {
   def getConnector(
       userProvidedDriverClass: Option[String],
       url: String,
-      userName: String): Connection = {
+      userName: String,
+      dbcp2Configs: String): Connection = {
     log.debug(s"${userProvidedDriverClass.getOrElse("")} $url $userName password")
     val subprotocol = new URI(url.stripPrefix("jdbc:")).getScheme
     val driverClass: Class[Driver] = getDriverClass(subprotocol, userProvidedDriverClass)
-    registerDriver(driverClass.getCanonicalName)
-    DriverManager.getConnection(url, userName, "password")
+
+    connectionPools.get(userName) match {
+      case Some(d) =>
+        d.getConnection
+      case None =>
+        val datasource = new BasicDataSource
+        datasource.setDriverClassName(driverClass.getCanonicalName)
+        datasource.setUrl(url)
+        datasource.setUsername(userName)
+        // for hdp older version support without dbcp2 configurations
+        if(dbcp2Configs == null) {
+            datasource.setInitialSize(2)
+            datasource.setMaxConnLifetimeMillis(2000)
+            datasource.setMaxTotal(100)
+            datasource.setMaxIdle(50)
```

Comment:
Based on the document, it said "But when you change maxIdle and make it lower than maxTotal without starting the evictor thread, it doesn't mean that your connections will not be closed. It means they will be closed immediately after releasing, with no delay, until their count not drop down to maxIdle." Therefore, if we set the (MaxTotal - MaxIdle) too big, the connection pool will remove some connections frequently. therefore, I suggest to MaxIdle not too small. Therefore, we can set the MaxTotal to 20, and MaxIdle to 10.